core/src/p2p.rs

```diff
@@ -0,0 +1,741 @@
+            .connect_peer_target
+            .lock();
+        target.clear();
+        target.extend_from_slice(new_target);
+        if let Some(index) = target
+            .iter()
+            .position(|peer_id| peer_id.public_key == self.public_key)
+        {
+            target.remove(index);
+        }
+        let target = target.clone();
+        let mut to_disconnect_keys = Vec::new();
+        for public_key in self
+            .connected_to_peers
+            .lock()
+            .keys()
+        {
+            if !target.iter().any(|id| id.public_key == *public_key) {
+                to_disconnect_keys.push(public_key.clone());
+            }
+        }
+        self.disconnect_peers(to_disconnect_keys)
+    }
+
+    #[allow(clippy::cognitive_complexity)]
+    /// The main loop for the listening thread.
+    fn listen_loop(&self, mut shutdown_receiver: tokio::sync::oneshot::Receiver<()>) {
+        let listener =
+            TcpListener::bind(&self.listen_addr).expect("Could not bind p2p tcp listener.");
+        listener
+            .set_nonblocking(true)
+            .expect("P2P subsystem could not enable nonblocking on listening tcp port.");
+
+        let mut last_connection_check = Instant::now();
+        loop {
+            // We have no obligations to network delivery so we simply exit on shutdown signal.
+            if shutdown_receiver.try_recv().is_ok() {
+                info!("P2P listen thread is being shut down");
+                break;
+            }
+            std::thread::sleep(Duration::from_millis((rand::random::<u64>() % 10) + 10));
+            self.check_connection(&mut last_connection_check);
+
+            let stream = match self.establish_new_connection(&listener) {
+                Ok(new_con) => new_con,
+                Err(e) => {
+                    debug!(?e);
+                    continue;
+                }
+            };
+
+            let (stream, crypto, other) = match self.perform_handshake(stream) {
+                Ok(tuple) => tuple,
+                Err(e) => {
+                    debug!(?e);
+                    continue;
+                }
+            };
+
+            if !self
+                .connect_peer_target
+                .lock()
+                .iter()
+                .any(|peer_id| peer_id.public_key == other)
+            {
+                trace!(%other, "Dropping because not in target.");
+                continue;
+            }
+
+            let mut connected_to_peers = self
+                .connected_to_peers
+                .lock();
+
+            if connected_to_peers.keys().any(|key| *key == other) {
+                trace!(%other, "Dropping because already connected.");
+            } else if let Err(e) =
+                finish_connection(crypto, stream, &mut connected_to_peers, &other)
+            {
+                error!(?e,
+                    %other,
+                    "Connecting to peer failed in final step."
```

Comment:
```suggestion
                    "Connection to peer failed in the final step."
```

what is the final step btw? can we be more specific why the connection failed?